### ISSUE TITLE
[gdb][libiconv] Fails to preview strings on UTF-8 locales

### ISSUE THREAD

**Issue by mcendu** [msys2/MINGW-packages#14947]
### Description / Steps to reproduce the issue

GDB fails to display strings on UTF-8 locales, that is, when "Beta: Use Unicode UTF-8 for worldwide language support" is checked on the host.

Requirements for reproducing:
- No `.gdbinit`.

Steps to reproduce:

1. Write a variant of Hello World as follows:
```c
#include <stdio.h>

int hello(char *str)
{
    return printf("%s\n", str);
}

int main(void)
{
    hello("Hello World!");
}
```

2. Compile with `cc -O0`.
3. Run `gdb a.exe` (or whatever you named in `-o).
4. Set a breakpoint at the line 5, then run.

This is the result of autodetection from GDB:

```
(gdb) show target-charset
The target character set is "auto; currently CP65001".
(gdb) show host-charset
The host character set is "auto; currently CP65001".
```

### Expected behavior

gdb should output `"Hello World!"` for `str` without any configuration.

### Actual behavior

In place of `"Hello World!"`, gdb outputs `<error reading variable: Converting character sets: Invalid argument.>` for the parameter `str`.

### Verification

- [X] I have verified that my MSYS2 is up-to-date before submitting the report (see https://www.msys2.org/docs/updating/)

### Windows Version

MINGW64_NT-10.0-22621

### MINGW environments affected

- [X] MINGW64
- [ ] MINGW32
- [ ] UCRT64
- [ ] CLANG64
- [ ] CLANG32
- [ ] CLANGARM64

### Are you willing to submit a PR?

Maybe

**Comment by mcendu** [comment]
From the source code: The list of encodings in GDB is generated from `iconv -l`. A possible upstream solution may be to uncomment [line 62 from libiconv](https://git.savannah.gnu.org/gitweb/?p=libiconv.git;a=blob;f=lib/encodings.def;h=41d8063e6b1997c5ac30f05e128d9b7bb5c40cde;hb=HEAD#l62):

```diff
@@ -60,5 +60,5 @@
 DEFENCODING(( "UTF-8",                  /* IANA, RFC 2279 */
             /*"UTF8",                      JDK 1.1 */
-            /*"CP65001",                   Windows */
+              "CP65001",                /* Windows */
             ),
             utf8,
```

**Comment by lhmouse** [comment]
```c
int printf(const char*, ...);

int hello(char *str)
{
    return printf("%s\n", str);
}

int main()
{
    hello("Hello World! テスト文字");
}
```

I can see the error, but despite that, outputs are not affected:

```text

C:\Users\lh_mouse\Desktop>gdb a
GNU gdb (GDB) 13.0.90.20230106-git
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-w64-mingw32".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from a...

warning: could not convert 'main' from the host encoding (CP65001) to UTF-32.
This normally should not happen, please file a bug report.
(gdb) r
Starting program: C:\Users\lh_mouse\Desktop\a.exe
[New Thread 992.0xa00]
[New Thread 992.0x1d24]
[New Thread 992.0x1ae4]
Hello World! テスト文字
[Thread 992.0x9c8 exited with code 0]
[Thread 992.0xa00 exited with code 0]
[Thread 992.0x1d24 exited with code 0]
[Inferior 1 (process 992) exited normally]
(gdb) quit

C:\Users\lh_mouse\Desktop>chcp
Active code page: 65001
```

Bad news is that, once UTF-8 is enabled for Windows, the conventional DBCS code pages (936, 932, etc.) are no longer valid.

**Comment by lhmouse** [comment]
The 32-bit GDB (linked against MSVCRT) crashes while loading the executable; not sure whether it's related to this issue:

```text
(gdb) file a.exe
Reading symbols from a.exe...
terminate called after throwing an instance of 'gdb_exception_error'
```

**Comment by lhmouse** [comment]
> From the source code: The list of encodings in GDB is generated from `iconv -l`. A possible upstream solution may be to uncomment [line 62 from libiconv](https://git.savannah.gnu.org/gitweb/?p=libiconv.git;a=blob;f=lib/encodings.def;h=41d8063e6b1997c5ac30f05e128d9b7bb5c40cde;hb=HEAD#l62):
> 
> ```diff
> @@ -60,5 +60,5 @@
>  DEFENCODING(( "UTF-8",                  /* IANA, RFC 2279 */
>              /*"UTF8",                      JDK 1.1 */
> -            /*"CP65001",                   Windows */
> +              "CP65001",                /* Windows */
>              ),
>              utf8,
> ```

This does not seem to work for me. After rebuilding and installing libiconv, `iconv -l | fgrep 65001` still results in nothing, and GDB still has the issue.

**Comment by lb90** [comment]
Should we add a manifest to `gdb.exe`? https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page#examples

**Comment by mcendu** [comment]
> > From the source code: The list of encodings in GDB is generated from `iconv -l`. A possible upstream solution may be to uncomment [line 62 from libiconv](https://git.savannah.gnu.org/gitweb/?p=libiconv.git;a=blob;f=lib/encodings.def;h=41d8063e6b1997c5ac30f05e128d9b7bb5c40cde;hb=HEAD#l62):
> > ```diff
> > @@ -60,5 +60,5 @@
> >  DEFENCODING(( "UTF-8",                  /* IANA, RFC 2279 */
> >              /*"UTF8",                      JDK 1.1 */
> > -            /*"CP65001",                   Windows */
> > +              "CP65001",                /* Windows */
> >              ),
> >              utf8,
> > ```
> 
> This does not seem to work for me. After rebuilding and installing libiconv, `iconv -l | fgrep 65001` still results in nothing, and GDB still has the issue.

Maybe try running `make -f Makefile.devel lib/aliases.h`? This could make it work.

**Comment by mcendu** [comment]
P.S. Should we consolidate changes to autogenerated files in one patch, or ditch them in favor of making `Makefile.devel` at our end? This is all making new patches difficult.

**Comment by alvinhochun** [comment]
I think upstream libiconv probably has a reason to not include CP65001 in the list. libiconv seems to have some sort of an alias for CP65001 here: https://git.savannah.gnu.org/gitweb/?p=libiconv.git;a=blob;f=libcharset/lib/localcharset.c;h=dbcd38468ddfc8f7c628f0e11c60b2481f8d28ff;hb=HEAD#l376

This file appears to originate from gnulib: https://git.savannah.gnu.org/gitweb/?p=gnulib.git;a=blob;f=lib/localcharset.c;h=7ed9c957db684186553b75decd2fb1e98c5292c2;hb=HEAD#l376

GCC appears to have its own [*extremely outdated copy* of it](https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=intl/localcharset.c;h=8ece6e39f69abf76922deb68bcbbce1331600c13;hb=HEAD), in fact all the way back from 2003, from [commit 360f8a2618bb34181bf5e298b291fc40955d07a0 of gnulib](https://git.savannah.gnu.org/gitweb/?p=gnulib.git;a=blob;f=lib/localcharset.c;h=d04d05368f96b53fa19b61ea254dee638920a69c;hb=360f8a2618bb34181bf5e298b291fc40955d07a0).

Would updating that file in GCC fix it properly?

**Comment by alvinhochun** [comment]
> Should we add a manifest to `gdb.exe`? https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page#examples
> 
> Note that the manifest only affects the A Win32 API, which that way expect UTF-8, convert to UTF-16 internally and call the corresponding W Win32 API. To have the CRT narrow functions expect UTF-8, an application has to call `setlocale(LC_ALL, ".utf8")`: https://developercommunity.visualstudio.com/t/create-a-utf8-c-runtime/351432. Of course that's only for the UCRT, not MSVCRT

In the context of this report, doing this would only force the bug to be reproducible for anyone on Windows 1903 or later.

Also there are a couple of other issues this will cause off the top of my head:

- MSVCRT does not really work correctly with UTF-8 code page.
- IIRC `OutputDebugString` calls from the debuggee process passes the debug output to the debugger *using the current ANSI code page*, unless the debugger opts in to receive the debug output as Unicode (wchar_t, i.e. UTF-16) using a special API. From past experience I don't think GCC is doing this, so if you force GCC to use UTF-8 code page but the debuggee doesn't, you will end up getting mojibake in GCC. (Granted, the reverse is also currently broken, i.e. GCC uses ANSI code page while debuggee uses UTF-8 code page.)

**Comment by mcendu** [comment]
Attempted a fix with #15030.

**Comment by mcendu** [comment]
Unlinked #15030 as it won't fix this issue without rebuilding GDB.
